SQL-1169: adds to_binary support for SQLUNKNOWN types

terakilobyte opened the pull request:
I chose to try to get the most accurate representation I could, but it may not be necessary.

terakilobyte commented:
Clippy failures didn't show for me, but @bucaojit is working on them in another PR #74

jchemburkar reviewed (commented):
Lgtm, pending the one question

terakilobyte commented on odbc/src/api/data.rs:
```diff
@@ -115,6 +115,21 @@ impl IntoCData for Bson {
+            Bson::JavaScriptCode(j) => Ok(j.into_bytes()),
+            Bson::JavaScriptCodeWithScope(j) => Ok(j.to_string().into_bytes()),
+            Bson::Symbol(s) => Ok(s.into_bytes()),
+            Bson::MinKey => Ok(f64::INFINITY.to_string().into_bytes()),
```
😆 Yes! I transposed them.

terakilobyte commented:
PR is now rebased to benefit from the clippy fixes. Additionally, I updated some of the returned information so that it might be more helpful to users after transformation.

bucaojit commented on odbc/src/api/data.rs:
```diff
@@ -115,6 +115,22 @@ impl IntoCData for Bson {
+            Bson::Undefined => Ok(char::from(0).to_string().into_bytes()),
+            Bson::Null => Ok(char::from(0).to_string().into_bytes()),
+            Bson::RegularExpression(r) => Ok(r.to_string().into_bytes()),
+            Bson::DbPointer(_) => Ok(char::from(0).to_string().into_bytes()),
```
For `DbPointer` there is a value we can display.  See if something like this works:
```
Bson::DbPointer(d) => Ok(serde_json::to_string(&d).unwrap().into_bytes()),
```

bucaojit commented on odbc/src/api/data.rs:
```diff
@@ -115,6 +115,22 @@ impl IntoCData for Bson {
+            Bson::Symbol(s) => Ok(s.into_bytes()),
+            Bson::MinKey => Ok("MinKey".to_string().into_bytes()),
+            Bson::MaxKey => Ok("MaxKey".to_string().into_bytes()),
             o => Err(ODBCError::RestrictedDataType(o.to_type_str(), BINARY)),
```
Looks like we're missing the [TIMESTAMP](https://github.com/mongodb/mongo-odbc-driver/blob/6073ee840d275fd96a0eba229a60f0c642d87faf/core/src/bson_type_info.rs#L317) value.

bucaojit commented on odbc/src/api/data.rs:
```diff
@@ -115,6 +115,22 @@ impl IntoCData for Bson {
             Bson::Binary(b) => Ok(b.bytes),
             Bson::Decimal128(d) => Ok(d.bytes().to_vec()),
+            Bson::Document(d) => Ok(d.to_string().into_bytes()),
+            Bson::Array(a) => Ok(serde_json::to_string(
```
Nice update to `Array`, it now shows as an array: `["a","b","c"]` instead of `abc`

bucaojit commented on odbc/src/api/data.rs:
```diff
@@ -115,6 +115,22 @@ impl IntoCData for Bson {
+            .unwrap()
+            .into_bytes()),
+            Bson::ObjectId(o) => Ok(o.to_string().into_bytes()),
+            Bson::Undefined => Ok(char::from(0).to_string().into_bytes()),
+            Bson::Null => Ok(char::from(0).to_string().into_bytes()),
```
I'm not entirely sure about representing these as `0`.  Since we return MinKey and MaxKey as their name, it might make sense to do the same.  
@nbagnard what would be a good way to represent/display the Null and Undefined values here?

nbagnard replied:
This is taken care of one step before in format_bson_data : Bson::Null and Bson::Undefined will show NULL (as in data is Null). For Bson::Null that because SQL does not make the distinction between Null and missing. For undefined, it's as good a representation as any ;)

bucaojit reviewed (changes requested):
Great work adding the binary support.  That definitely helps us understand what other changes are needed with Power BI.

bucaojit commented on odbc/src/api/data_tests.rs:
```diff
@@ -761,26 +777,13 @@ mod unit {
+                bin_val_test(
+                    DOC_COL,
+                    &[
+                        123, 32, 34, 120, 34, 58, 32, 52, 50, 44, 32, 34, 121, 34, 58, 32, 52, 50,
```
I have one suggestion for these tests.  Now that we have binary values that are readable.  How about we modify these tests to have the expected value be the string with .as_bytes().  That would make it more readable, such as: `"{ \"x\": 42, \"y\": 42 }".as_bytes()".  See if it works, otherwise we can stick with how it is written.

terakilobyte replied:
great suggestion, I'll work on it

nbagnard reviewed (changes requested):
It's looking good!

Can we make sure that the string representation is the same for the C_binary and C_[W]CHAR? If it looks like this is not as straightforward as it sounds, let me know and we can see if it is better to create a separate PR.

Adding a test making sure that the string representation of the data for each bson type is indeed the expected relaxed extended json would also be great. Not sure if we could handle that directly with the test harness, but that would be the easiest. Something similar to how we check the values returned for the collection types_other on the JDBC driver.
This can also be a separate PR to keep things moving, especially since Oliver already gave you a suggestion to get things more easily human-readable.

terakilobyte commented on odbc/src/api/data.rs:
```diff
@@ -115,7 +114,23 @@ impl IntoCData for Bson {
             Bson::Int64(i) => Ok(i.to_le_bytes().to_vec()),
             Bson::Binary(b) => Ok(b.bytes),
             Bson::Decimal128(d) => Ok(d.bytes().to_vec()),
```
As of now, we'll just send the bytes back. I think it may be better to wait for the full implementation of this type before we try to represent it concretely. Thoughts?

nbagnard replied:
Let's return the json represensation. It will make the data usable until the Rust team has Decimal128 supported "natively" in their driver

terakilobyte commented on odbc/src/api/data.rs:
```diff
@@ -115,7 +114,23 @@ impl IntoCData for Bson {
+            )
+            .unwrap()
+            .into_bytes()),
+            Bson::ObjectId(_) => Ok(self.to_json().into_bytes()),
```
simplified even more

terakilobyte commented on odbc/src/api/data.rs:
```diff
@@ -115,7 +114,23 @@ impl IntoCData for Bson {
             Bson::Binary(b) => Ok(b.bytes),
             Bson::Decimal128(d) => Ok(d.bytes().to_vec()),
-            o => Err(ODBCError::RestrictedDataType(o.to_type_str(), BINARY)),
+            Bson::Document(d) => Ok(d.to_string().into_bytes()),
```
It doesn't match in Tableau. For example, ObjectID by default shows as extjson `{"$oid": "abcd"}`, but `_id::string` just shows the string value "abcd". I'm happy to make them the same though!

I've updated the other tests to show what the string repr looks like for each type.

nbagnard replied:
As discussed in one-on-one, changing all types mapped to call to_json() doe the trick 👍

terakilobyte commented:
Turns out, by eliminating match arms I was able to pass all of the unknown types (Decimal128 excluded) directly to `to_json` to get the extjson repr.

bucaojit commented on odbc/src/api/data_tests.rs:
```diff
@@ -789,12 +784,33 @@ mod unit {
+                bin_val_test(JS_W_S_COL, "{\"$code\":\"log(\\\"hello\\\" + x + \\\"world\\\")\",\"$scope\":{\"x\":{\"$numberInt\":\"42\"}}}".as_bytes(), SqlReturn::SUCCESS);
+                bin_val_test(
+                    MAXKEY_COL,
+                    &[123, 34, 36, 109, 97, 120, 75, 101, 121, 34, 58, 49, 125],
```
Nice work on the test cases.  For MaxKey and MinKey we should also be able to represent them as:
"{\"$maxKey\":1}".as_bytes()
"{\"$minKey\":1}".as_bytes()

bucaojit reviewed (approved):
LGTM.
That's cool how the branches were able to be simplified. 

I had one small test suggestion, no need to retag.

nbagnard reviewed (approved):
LGTM! Nice work :)

terakilobyte commented:
evergreen merge
